pkg/status/combinedstatus-resolution.go

```diff
@@ -0,0 +1,308 @@
+	ownerReference *metav1.OwnerReference
+}
+
+// statusCollectorData is a struct that represents the data of a status
+// collector in a combinedstatus resolution.
+type statusCollectorData struct {
+	*statusCollectorClauses
+
+	// workStatusRefToData is a map of workstatus references to their
+	// corresponding work status data.
+	// A workstatus reference is identical to its source object's identifier,
+	// apart from the namespace, which is set to the associated clusterName.
+	// If a workstatus reference is mapped to nil, it means that the workstatus
+	// has not been evaluated against the statuscollector's clauses.
+	workStatusRefToData map[util.ObjectIdentifier]workStatusData
+}
+
+// workStatusData is a struct that represents the evaluation of a workstatus
+// against a statuscollector's clauses.
+type workStatusData struct {
+	// filterEval is a boolean that represents the evaluation of the filter
+	// expression against the workstatus. If true, the workstatus is included
+	// in the statuscollector's result.
+	filterEval bool
+
+	// groupByEval is a map of groupBy expression names to their evaluated values.
+	groupByEval map[string]any
+	// combinedFieldsEval is a map of combinedFields expression names to their
+	// evaluated values. CombinedField types:
+	// - COUNT: the number of workstatuses per groupBy value. If groupBy is
+	//   empty, the count is the number of workstatuses.
+	// - SUM: the sum of the values of the workstatuses groupBy values.
+	// - AVG: the average of the values of the workstatuses groupBy values.
+	// - MIN: the minimum value of the workstatuses groupBy values.
+	// - MAX: the maximum value of the workstatuses groupBy values.
+	combinedFieldsEval map[string]any
+
+	selectEval map[string]any
+}
+
+// setStatusCollectors sets ALL the statuscollectors relevant to the
+// combinedstatus resolution.
+// The given map is expected to be thread-safe, its values contain valid
+// expressions, and are immutable. If a value is nil, it is assumed that the
+// associated statuscollector has not been processed.
+// The function returns a tuple (bool1, bool2):
+//
+// - bool1: true if one or more statuscollectors were removed.
+//
+// - bool2: true if one or more statuscollectors were added.
+func (c *combinedStatusResolution) setStatusCollectors(statusCollectorNameToClauses map[string]*statusCollectorClauses) (bool, bool) {
+	c.Lock()
+	defer c.Unlock()
+
+	workStatusRemoved, workStatusAdded := false, false
+
+	// remove statuscollector data that are not relevant anymore and update the
+	// statuscollector data that are. If one of the latter is updated, add its
+	// workstatuses to be queued up.
+	for statusCollectorName, statusCollectorData := range c.statusCollectorNameToData {
+		clauses, ok := statusCollectorNameToClauses[statusCollectorName]
+		if !ok {
+			delete(c.statusCollectorNameToData, statusCollectorName)
+			workStatusRemoved = true
+			continue
+		}
+
+		if !statusCollectorData.equals(clauses) {
```

Comment:
This reads oddly, it is asking whether a statusCollectorData _equals_ a clauses. But you mean to ask whether the _clauses of_ the statusCollectorData equals another clauses. It would be better if this were more obvious here and at the other place like it.